### ISSUE TITLE
stock status can't update correctly after  product stocks sync through API

### ISSUE THREAD

**Issue by missvanna** [woothemes/woocommerce#32047]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

stock status can't update correctly after product stocks sync through API
I see this:
Out of stock (8)
Out of stock (3)
etc..

my code as below:

```php
function synchronize_hktvmall_inventory( ) {
        global $wpdb;
        $sql  = $wpdb->prepare( "SELECT count(pm.post_id) FROM $wpdb->posts as p left join $wpdb->postmeta as pm on pm.post_id = p.ID WHERE post_status like 'publish' and post_type LIKE 'product%' and pm.meta_key like '_sku'");
        $total_products_count  = $wpdb->get_var($sql);
        $size = 80;
        $times = ceil($total_products_count/$size);
        

        for ($index=0; $index < $times; $index++) {
                $start = $size * $index;
                $sql  = $wpdb->prepare( "SELECT pm.meta_value FROM $wpdb->posts as p left join $wpdb->postmeta as pm on pm.post_id = p.ID WHERE post_status like 'publish' and post_type LIKE 'product%' and pm.meta_key like '_sku' LIMIT $start, $size");
                $products_results  = $wpdb->get_results($sql);
                $product_codes_arr = [];
                $hktvmall_code_prefix = 'H7988001_S_';
                foreach ( $products_results as $product ) {
                        $product_code_arr = array(
                                "warehouseId" => "H798800101",
                                "productId" => $hktvmall_code_prefix.$product->meta_value,
                        );
                        array_push($product_codes_arr,$product_code_arr);
                }
                $data = json_encode($product_codes_arr);
                $payload = array(
                        "sub" => "shoalter",
                        "name" => "shoalter",
                        "iat" => time(),
                        "x-api-key" => "dcda80cf-1355-4493-a7ce-45ab1c0cfe68"
                );
                $jwt = JWT::encode($payload, PRIVATE_KEY, 'RS256');
                //get hktvmall stock 
                $result = get_hktvmall_stock($jwt,$data);
                if ($result["code"] == "success") {
                        foreach ($result["data"] as $info ) {
                                if ($info["status"] !== "failed") {
                                        $stock = $info["availableToSell"];
                                        $hktvmall_product_code = $info["productId"];
                                        $product_code = substr($hktvmall_product_code,11);
                                        $sql  = $wpdb->prepare( "SELECT post_id, p.post_type, p.post_parent FROM $wpdb->posts as p left join $wpdb->postmeta as pm on pm.post_id = p.ID WHERE meta_key like '_sku' and meta_value like '$product_code'");
                                        $row  = $wpdb->get_row($sql);
                                        if ( is_object( $row ) && $stock >=0 ) {
                                                $post_id = $row->post_id;
                                                $post_parent = $row->post_parent;
                                                $post_type = $row->post_type;
                                                $wpdb->query( $wpdb->prepare( "UPDATE {$wpdb->postmeta} SET meta_value = %s WHERE post_id = %s and meta_key like '_stock'" , $stock, $post_id ) );
                                                if ($stock >0) {
                                                        $wpdb->query( $wpdb->prepare( "UPDATE {$wpdb->postmeta} SET meta_value = 'instock' WHERE post_id = %s and meta_key = '_stock_status' and meta_value like 'onbackorder'" , $post_id ) );
                                                        if ($post_type == 'product_variation' && $post_parent > 0) {
                                                                $wpdb->query( $wpdb->prepare( "UPDATE {$wpdb->postmeta} SET meta_value = 'instock' WHERE post_id = %s and meta_key = '_stock_status' and meta_value like 'onbackorder'" , $post_parent ) );
                                                        }
                                                } 
                                        }                             
                                }
                        }
                }

        }
    
}

add_action( 'synchronize_hktvmall_inventory', 'synchronize_hktvmall_inventory' );
```
### Expected behavior

hope stock status can update correctly

### Actual behavior

stock status can't update correctly

### Steps to reproduce

<img width="777" alt="image" src="https://user-images.githubusercontent.com/101256367/157426757-ab90228e-67e1-495b-bdbc-1ab3502ab2d4.png">

### WordPress Environment

WooCommerce Version: 6.2.0
WordPress Version: 5.9.1

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [ ] I can reproduce this bug consistently using the steps above.

**Comment by claudiosanches** [comment]
Hi @missvanna,

While our goal is to address all the issues reported in this repository, GitHub should be treated as a place to report confirmed bugs only.

The type of issue you submitted looks like a custom code related question. Due to the number of bug reports being submitted, we are prioritizing work on them and are not able to attend to custom code questions. 

The best place to get help further is in the [WooCommerce Community Slack](https://woocommerce.com/community-slack/) in the `#developers` channel.

The code you provided is making requests to WordPress database directly without any WooCommerce iteration, WooCommerce provides a REST API and also a full PHP API to properly update and sync product stock.

Here an example using the [`wc_update_product_stock()`](https://woocommerce.github.io/code-reference/files/woocommerce-includes-wc-stock-functions.html) function and cleaning the transients later with [`wc_delete_product_transients()`](https://woocommerce.github.io/code-reference/namespaces/default.html#function_wc_delete_product_transients).

```php
$product = wc_get_product( $product_id );  
wc_update_product_stock( $product,  $qty_new , 'set' );
wc_delete_product_transients( $product_id ); // Cleaning product transients will help to display the correct value fast
```

If you confirm the bug, please open a new issue and provide us with clear steps to reproduce it.

I am closing this issue for now. If I missed something, please leave a comment and we will take a second look.

**Comment by woocommercebot** [comment]
Hi @missvanna,

  While our goal is to address all the issues reported in this repository,   GitHub should be treated as a place to report confirmed bugs only.

  The type of issue you submitted looks like a support request which may or may not reveal a bug once proper   troubleshooting is done.  In order to confirm the bug, please follow one of the steps below:

  - Review [WooCommerce Self-Service Guide](https://docs.woocommerce.com/document/woocommerce-self-service-guide/)   to see if the solutions listed there apply to your case;
  - If you are a paying customer of WooCommerce, contact WooCommerce support by   [opening a ticket or starting a live chat](https://woocommerce.com/contact-us/);
  - Make a post on [WooCommerce community forum](https://wordpress.org/support/plugin/woocommerce/)

  If you confirm the bug, please provide us with clear steps to reproduce it.

  We are closing this issue for now as it seems to be a support request and not a bug.   If we missed something, please leave a comment and we will take a second look.
